### ISSUE TITLE
Range header for 206 Partial Content is not respected for binary media requests

### ISSUE THREAD

**Issue by ThaNarie** [kettanaito/msw#1972]
### Prerequisites

- [X] I confirm my issue is not in the [opened issues](https://github.com/mswjs/msw/issues)
- [X] I confirm the [Frequently Asked Questions](https://mswjs.io/docs/faq) didn't contain the answer to my issue

### Environment check

- [X] I'm using the [latest](https://github.com/mswjs/msw/releases/latest) `msw` version
- [X] I'm using Node.js version 14 or higher

### Reproduction repository

https://codesandbox.io/p/devbox/msw-206-partial-content-t9m6r3

### Reproduction steps

`npm run start`

---

Note that the linked CodeSandbox template is using 1.x, and doesn't work due to SSL errors by default. I ended up forking the Svelte example instead.

### Current behavior

When seeking in media, a `Range` request is sent to only request the relevant bytes. This header doesn't seem to be forwarded by the fetch request in the Service Worker, so the complete binary file is downloaded every time you seek.

### Expected behavior

The `Range` header to be forwarded so that the response is a correct **206 Partial Content**.

---

Not sure if this is a bug, or simply a missing feature, but for now I have to explicitly disable `.mp4` requests in my service-worker file because otherwise seeking in video is unworkable (having to wait 10+ seconds instead of <1 second).

If not fixable, or at least not short term, it would be useful to have this documented as a missing feature.

**Comment by kettanaito** [comment]
Hi, @ThaNarie. Thanks for raising this. 

It looks like the Service Worker [intentionally omits the `Range` request header](https://web.dev/articles/sw-range-requests). Instead, it encourages the fetch listener to return early if it detects a `Range` header on the request. 

The article above illustrates the issue you're having. 

This means that in order to let the worker handle ranged responses correctly, we must also add an early return for such requests before calling `event.respondWith()`:

https://github.com/mswjs/msw/blob/62a5574af5f048566911518bd8b7ba0f254eb72b/src/mockServiceWorker.js#L111

This basically means that we need to choose either of these:

1. Requests with a `Range` header _will never be mocked_. The handler lookup is async and it's happening in the `event.respondWith()` scope because that's the only scope in the fetch event listener that's allowed to be async (fetch event listeners are synchronous). 
2. Requests with a `Range` header can be mocked but if they have no request handlers (are bypassed), they won't forward the `Range` header to the original destination. This means having the issue you are describing as an actual behavior/limitation. 

I need to read more about why the `Range` request header is not forwarded when we do `fetch(event.request)`. Sounds like an explicit spec limitation.

**Comment by kettanaito** [comment]
Okay, taking a step back a bit. I misread the article. It clearly states that the `Range` header is not forwarded in the requests in a service worker _in browsers that behave incorrectly_. Later into the article, it illustrates that this code should just work:

```js
self.addEventListener('fetch', (event) => {
  // The Range: header will pass through in browsers
  // that behave correctly.
  event.respondWith(fetch(event.request));
});
```

**Comment by kettanaito** [comment]
The first observable difference I can see is that requests with a worker enabled don't have the upper boundary on the range:

```
With SW:
  Range: bytes=40304640-

No SW: 
  Range: bytes=51314688-92274687
```

I'm not proficient enough to know how exactly the server will treat a `Range` request without the upper boundary. Can it be that the lack of causes the server to download not a fixed chunk but a chink _starting_ from the lower boundary and to the end of the video? 

This seems to be the case based on MDN:

> [<range-end>](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range#range-end)
An integer in the given unit indicating the end position of the requested range. This value is optional and, if omitted, the end of the resource is used as the end of the range.
> — [Source](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range#range-end)

**Comment by kettanaito** [comment]
@ThaNarie, I've just added isolated browser tests for 206 responses in https://github.com/mswjs/msw/pull/1975. It passes when forwarding the request (your use case) and also mocking a request with a `Range` header. 

As of now, I conclude this issue to be someplace else than MSW. Feel free to experiment with the test cases and let me know if you get to reproduce the issue. Thanks.

**Comment by kettanaito** [comment]
## Released: v2.1.3 🎉

This has been released in v2.1.3!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.1.3)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.1.3)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
